Constrexpr.Numeral without bigint

letouzey opened the pull request:
This is an early part of PR #496 (Numeral Notation), the one that deals with internal layers of Coq.
To allow parsing/printing of numbers without relying on the ML bigint.ml library, I propose to keep `Constrexpr.Numeral` in a more primitive form (a string of the decimal digits of the number, plus a boolean sign).

With this PR, the code of PR #496 could live in a separate plugin (and opam package) until the remaining issues are fixed and Numeral Notation is ready for full integration.

As mentioned by @JasonGross in #496, ssr will need an overlay with this code.

JasonGross commented:
I'm not sure that this is quite enough for a bug-free opam package; in particular, I worry that `declare_rawnumeral_interpreter` may in its present form be incompatible with backtracking.  However, I am strongly in favor of this getting into 8.7, even if we have to bug-fix it later.

In fact, I'd advocate getting all of #496 into 8.7, possibly with the exception of the numeral notations for `int_scope` and `uint_scope` (which are what break math-comp and iris-coq), if you can't fix the locality issues in time.

ejgallego commented:
The backtracking problem is already present, and possible to fix with a bit of hacking, but it would require a good understanding on what is the best way to recreate `prim_token_interpreter_tab`.

letouzey commented:
Guys, what "backtracking problem" are you speaking of ? I'd be glad to have an example to work on. I'm also seriously dubious about that, since these commits are strictly about changing representation (a `string*bool` instead of a `Bigint.bigint`). If a problem exists with my new `declare_rawnumeral_interpreter`, I do not see how the original `declare_numeral_interpreter` could be immune of it, even though it's the current way numbers are parsed/printed today.

ejgallego commented:
Both imperative objects `prim_token_interpreter_tab` and `prim_token_key_table` are not registered in the summary, thus they won't be restored when backtracking.

JasonGross commented:
The backtracking problem is https://github.com/coq/coq/pull/496#issuecomment-305267813:
```bash
$ echo "Global Unset Universe Minimization ToSet.
Polymorphic Axiom nat_of_int'@{i} : Decimal.int -> option nat.
Polymorphic Axiom nat_to_int'@{i} : nat -> Decimal.int.
Delimit Scope nat'_scope with nat'.
Open Scope nat'_scope.
Check 0.
Numeral Notation nat nat_of_int' nat_to_int' : nat'_scope.
Backtrack 4 0 0.
Check 0.
" | ~/.local64/coq/coq-numerals/bin/coqtop -q -time -emacs
Welcome to Coq jgross-Leopard-WS:/home/jgross/Documents/repos/coq,pr/496 (92fad459806e27482909e82b2b89d2dc0641a61c)
<infomsg>Skipping rcfile loading.</infomsg>

<prompt>Coq < 1 || 0 < </prompt>Chars 0 - 41 [Global~Unset~Universe~Minimiza...] <infomsg>0. secs (0.u,0.s)</infomsg>

<prompt>Coq < 2 || 0 < </prompt>Chars 42 - 104 [Polymorphic~Axiom~(nat_of_int'...] <infomsg>nat_of_int' is declared</infomsg>
<infomsg>0. secs (0.u,0.s)</infomsg>

<prompt>Coq < 3 || 0 < </prompt>Chars 105 - 160 [Polymorphic~Axiom~(nat_to_int'...] <infomsg>nat_to_int' is declared</infomsg>
<infomsg>0. secs (0.u,0.s)</infomsg>

<prompt>Coq < 4 || 0 < </prompt>Chars 161 - 196 [Delimit~Scope~nat'_scope~with~...] <infomsg>0. secs (0.u,0.s)</infomsg>

<prompt>Coq < 5 || 0 < </prompt>Chars 197 - 219 [Open~Scope~nat'_scope.] <infomsg>0. secs (0.u,0.s)</infomsg>

<prompt>Coq < 6 || 0 < </prompt>Chars 220 - 228 [Check~0.] 0
     : nat
<infomsg>0. secs (0.u,0.s)</infomsg>

<prompt>Coq < 7 || 0 < </prompt>Chars 229 - 287 [Numeral~Notation~nat~nat_of_in...] <infomsg>0. secs (0.003u,0.s)</infomsg>

<prompt>Coq < 8 || 0 < </prompt>Chars 288 - 304 [Backtrack~4~0~0.]

<prompt>Coq < 4 || 0 < </prompt>Chars 305 - 313 [Check~0.] Segmentation fault
```
and
```bash
$ echo "Global Unset Universe Minimization ToSet.
Polymorphic Axiom nat_of_int'@{i} : Decimal.int -> option nat.
Polymorphic Axiom nat_to_int'@{i} : nat -> Decimal.int.
Delimit Scope nat'_scope with nat'.
Open Scope nat'_scope.
Check 0.
Numeral Notation nat nat_of_int' nat_to_int' : nat'_scope.
Backtrack 7 0 0.
Check 0.
" | ~/.local64/coq/coq-numerals/bin/coqtop -q -time -emacs
Welcome to Coq jgross-Leopard-WS:/home/jgross/Documents/repos/coq,pr/496 (92fad459806e27482909e82b2b89d2dc0641a61c)
<infomsg>Skipping rcfile loading.</infomsg>

<prompt>Coq < 1 || 0 < </prompt>Chars 0 - 41 [Global~Unset~Universe~Minimiza...] <infomsg>0. secs (0.u,0.s)</infomsg>

<prompt>Coq < 2 || 0 < </prompt>Chars 42 - 104 [Polymorphic~Axiom~(nat_of_int'...] <infomsg>nat_of_int' is declared</infomsg>
<infomsg>0. secs (0.u,0.004s)</infomsg>

<prompt>Coq < 3 || 0 < </prompt>Chars 105 - 160 [Polymorphic~Axiom~(nat_to_int'...] <infomsg>nat_to_int' is declared</infomsg>
<infomsg>0. secs (0.u,0.s)</infomsg>

<prompt>Coq < 4 || 0 < </prompt>Chars 161 - 196 [Delimit~Scope~nat'_scope~with~...] <infomsg>0. secs (0.u,0.s)</infomsg>

<prompt>Coq < 5 || 0 < </prompt>Chars 197 - 219 [Open~Scope~nat'_scope.] <infomsg>0. secs (0.u,0.s)</infomsg>

<prompt>Coq < 6 || 0 < </prompt>Chars 220 - 228 [Check~0.] 0
     : nat
<infomsg>0. secs (0.u,0.s)</infomsg>

<prompt>Coq < 7 || 0 < </prompt>Chars 229 - 287 [Numeral~Notation~nat~nat_of_in...] <infomsg>0. secs (0.u,0.s)</infomsg>

<prompt>Coq < 8 || 0 < </prompt>Chars 288 - 304 [Backtrack~7~0~0.]

<prompt>Coq < 7 || 0 < </prompt>Chars 305 - 313 [Check~0.] <infomsg>0. secs (0.u,0.s)</infomsg>
Toplevel input, characters 0-8:
> Check 0.
> ^^^^^^^^
Anomaly: Parsing virtual machine value expected universe level, got Vprod.
Please report at http://coq.inria.fr/bugs/.

<prompt>Coq < 7 || 0 < </prompt>
```

JasonGross commented:
I wouldn't be surprised if there is already a similar issue that can occur when you backtrack across a `Require`, but backtracking across `Require` is somewhat rare, whereas backtracking across adding a numeral notation is more common.

ejgallego commented:
> I wouldn't be surprised if there is already a similar issue

Indeed there is but it is hard to observe as many of the relevant modules are loaded at init time and can't be backtracked upon, unless you disable loading of prelude.

JasonGross commented:
> Indeed there is but it is hard to observe as many of the relevant modules are loaded at init time and can't be backtracked upon, unless you disable loading of prelude.

Even if I run with `-nois` and backtrack across loading of the prelude, I still can't reproduce it on present trunk.

ejgallego commented:
In this case what you can observe is these two tables being out of sync when you backtrack on a `Declare ML Module` , however I still don't understand why you get the segfault in the second case, it is not obvious from the code.

letouzey commented:
This backtracking issue is a separate topic, orthogonal to the code of this particular limited PR, and it shouldn't block its integration in 8.7. Of course the bug deserves fixing, but that could come as a bugfix during 8.6 beta.

ejgallego commented:
@letouzey indeed, sorry for the confusion, the segfault problem is not related to backtracking but to improper handling of polymorphic constants in #496 , we should move the discussion there.

JasonGross commented on ide/texmacspp.ml:
```diff
@@ -382,7 +382,7 @@ and pp_token ?loc tok =
     match tok with
     | String s -> PCData s
-    | Numeral n -> PCData (to_string n) in
+    | Numeral (n,s) -> PCData (if s then n else "- "^n) in
```
Shouldn't this be `"-"^n` for printing compatibility?  Also, you should probably strip leading zeros, no?

JasonGross commented on interp/constrexpr_ops.ml:
```diff
@@ -45,8 +45,11 @@ let names_of_local_binders bl =
 (* Functions on constr_expr *)
 
+(* Note: redundant Numeral representations such as -0 and +0 (or different
+   numbers of leading zeros) are considered different here. *)
```
Caring about differences in the number of leading zeros breaks compatibility with v8.6 and trunk, right?

JasonGross commented on interp/notation.ml:
```diff
@@ -440,8 +457,8 @@ let find_notation ntn sc =
 let notation_of_prim_token = function
-  | Numeral n when is_pos_or_zero n -> to_string n
-  | Numeral n -> "- "^(to_string (neg n))
+  | Numeral (n,true) -> n
```
You need to strip leading zeros here, or you'll break the test-suite.

JasonGross commented on interp/notation.ml:
```diff
@@ -440,8 +457,8 @@ let find_notation ntn sc =
-  | Numeral n when is_pos_or_zero n -> to_string n
-  | Numeral n -> "- "^(to_string (neg n))
+  | Numeral (n,true) -> n
+  | Numeral (n,false) -> "- "^n
```
You should to strip leading zeros here.

JasonGross commented on printing/ppconstr.ml:
```diff
@@ -231,7 +231,7 @@ let tag_var = tag Tag.variable
 
   let pr_prim_token = function
-    | Numeral n -> str (Bigint.to_string n)
+    | Numeral (n,s) -> str (if s then n else "- "^n)
```
This should be `"-"^n` for compatibility with trunk/v8.6.

maximedenes commented:
@letouzey Are you working on a fix, or should we postpone this to 8.8?

Zimmi48 commented:
@maximedenes Note that @JasonGross has prepared a fix for the initial PR (see #496).

letouzey commented:
Rebased

maximedenes commented:
I'm afraid we're gonna have to postpone this, unless a fix is pushed today.

letouzey commented:
Status update on this :
  - Rebased
  - Fixed the space before "-" mentionned by @JasonGross 
  - *Not* discarding leading zeros here, unlike suggested by @JasonGross. The philosophy is to leave the string coding the number as untouched as possible. The leading zeros will be handled by higher layer (see the full Numeral Notation PR #496). This might indeed lead to minor changes in behavior, in very particular situations such as `bugs/closed/3036.v` and its 00 numbers and its Qarith notation for just 0 and 1. So I've adapted `3036.v` (this bug had anyway nothing to see with zero being written 0 or 00).
  - With this minor adaptation, the test-suite is ok, as well as all other travis tests (except fiat something that timeouts)

Still to investigate (but *not* blocking the integration of this PR):
  - add synchronization for the atom imperative tables
  - universe issues with the full Numeral Notation PR

ppedrot commented:
LGTM.
